src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @emails react-core
+ */
+
+'use strict';
+
+describe('ReactDOM unknown attribute', () => {
+  var React;
+  var ReactDOM;
+
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    ReactDOM = require('react-dom');
+  });
+
+  function normalizeCodeLocInfo(str) {
+    return str && str.replace(/\(at .+?:\d+\)/g, '(at **)');
+  }
+
+  function testUnknownAttributeRemoval(givenValue) {
+    var el = document.createElement('div');
+    ReactDOM.render(<div unknown="something" />, el);
+    expect(el.firstChild.getAttribute('unknown')).toBe('something');
+    ReactDOM.render(<div unknown={givenValue} />, el);
+    expect(el.firstChild.hasAttribute('unknown')).toBe(false);
+  }
+
+  function testUnknownAttributeAssignment(givenValue, expectedDOMValue) {
+    var el = document.createElement('div');
+    ReactDOM.render(<div unknown="something" />, el);
+    expect(el.firstChild.getAttribute('unknown')).toBe('something');
+    ReactDOM.render(<div unknown={givenValue} />, el);
+    expect(el.firstChild.getAttribute('unknown')).toBe(expectedDOMValue);
+  }
+
+  describe('unknown attributes', () => {
+    it('removes values null and undefined', () => {
+      testUnknownAttributeRemoval(null);
+      testUnknownAttributeRemoval(undefined);
+    });
+
+    it('changes values true, false to null, and also warns once', () => {
+      spyOn(console, 'error');
+
+      testUnknownAttributeAssignment(true, null);
+      testUnknownAttributeAssignment(false, null);
+
+      expectDev(
+        normalizeCodeLocInfo(console.error.calls.argsFor(0)[0]),
+      ).toMatch(
+        'Warning: Received `true` for non-boolean attribute `unknown`. ' +
+          'If this is expected, cast the value to a string.\n' +
+          '    in div (at **)',
+      );
+      expectDev(console.error.calls.count()).toBe(1);
```

Comment:
It might be better to put this assertion first because otherwise the previous assertion would probably blow up with `Cannot read property 0 of undefined`. I haven't verified this though.